test/state/state.cpp

```diff
@@ -460,6 +480,89 @@ TransactionReceipt transition(const StateView& state_view, const BlockInfo& bloc
     assert(sender_acc.nonce < Account::NonceMax);  // Required for valid tx.
     ++sender_acc.nonce;                            // Bump sender nonce.
 
+    int64_t delegation_refund = 0;
+
+    for (const auto& auth : tx.authorization_list)
+    {
+        // 1. Verify the chain id is either 0 or the chain’s current ID.
+        // FIXME: Transaction chain id is not the chain id. Can it be 0?
+        if (auth.chain_id != 0 && auth.chain_id != tx.chain_id)
+            continue;
+
+        // 2. Verify the nonce is less than 2**64 - 1.
+        if (auth.nonce == Account::NonceMax)
+            continue;
+
+        // 3. Verify if the signer has been successfully recovered from the signature.
+        //    authority = ecrecover(...)
+        if (!auth.signer.has_value())
+            continue;
+
+        // s value must be less than or equal to secp256k1n/2, as specified in EIP-2.
+        if (auth.s > SECP256K1N_OVER_2)
+            continue;
+
+        // Get or create the authority account.
+        // It is still empty at this point until nonce bump followed by successful authorization.
+        auto& authority = state.get_or_insert(*auth.signer, {.erase_if_empty = true});
+
+        // 4. Add authority to accessed_addresses (as defined in EIP-2929.)
+        authority.access_status = EVMC_ACCESS_WARM;
+
+        // 5. Verify the code of authority is either empty or already delegated.
+        if (authority.code_hash != Account::EMPTY_CODE_HASH &&
+            !is_code_delegated(state.get_code(*auth.signer)))
+            continue;
+
+        // 6. Verify the nonce of authority is equal to nonce.
+        // In case authority does not exist in the trie, verify that nonce is equal to 0.
+        if (auth.nonce != authority.nonce)
+            continue;
+
+        // 7. Add PER_EMPTY_ACCOUNT_COST - PER_AUTH_BASE_COST gas to the global refund counter
+        // if authority exists in the trie.
+        // We are interested in _empty_ accounts (EIP-161) because _empty_ implies they don't exist
+        // in the state (EIP-7523). Successful authorisation validation makes an account non-empty.
+        // We apply the refund only if the account has been non-empty before.
+        // TODO: The additional condition (erase_if_empty) is to handle existent-but-empty which
```

Comment:
`authority` here doesn't necessary represent what is commited to the state, but intermediate in-memory account object, existing during transaction processing. It can indeed be empty, not only in testing environments (this is complaint about comment formulation)

`erase_if_empty` here seems to be a hacky way to represent the flag "account was created in this transaction". Instead you should have a separate flag that is set when authority is not found in the state the first time it is mentioned.

Reply:
The `erase_if_empty` means "empty account created in this transaction". The name is probably bad because it says what to do with it rather than what it is, but besides the name I think it suits this case pretty well.